File: day-07/part-2/ayoub.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func findRoot(graph map[string][]string) string {
	parent := make(map[string]string)

	for node, children := range graph {
		for _, child := range children {
			parent[child] = node
		}
	}

	for node := range graph {
		if _, ok := parent[node]; !ok {
			return node
		}
	}
	return ""
}

func recCalcTotalWeights(graph map[string][]string, weight map[string]int, root string, cache map[string]int) int {
	if val, exists := cache[root]; exists {
		return val
	}
	res := weight[root]
	for _, child := range graph[root] {
		cache[child] = recCalcTotalWeights(graph, weight, child, cache)
		res += cache[child]
	}
	cache[root] = res
	return cache[root]
}

func calcTotalWeights(graph map[string][]string, weight map[string]int, root string) map[string]int {
	res := make(map[string]int)
	for node := range graph {
		res[node] = recCalcTotalWeights(graph, weight, node, make(map[string]int))
	}
	return res
}

func isBalanced(graph map[string][]string, totalWeight map[string]int, root string) bool {
	last := -1
	for _, child := range graph[root] {
		if last == -1 {
			last = totalWeight[child]
		} else if last != totalWeight[child] {
			return false
		}
	}
	return true
}

func findCorrectedWeight(graph map[string][]string, weight map[string]int, totalWeight map[string]int, root string) int {
	if len(graph[root]) == 0 || isBalanced(graph, totalWeight, root) {
		return -1
	}

	// Look for the wrong weight
	wrongChild := ""
	wantedWeight := 0
	occurence := make(map[int]int)
	for _, child := range graph[root] {
		if _, exists := occurence[totalWeight[child]]; exists {
			occurence[totalWeight[child]]++
			wantedWeight = totalWeight[child]
		} else {
			occurence[totalWeight[child]] = 1
		}
	}
	for _, child := range graph[root] {
		if occurence[totalWeight[child]] == 1 {
			wrongChild = child
			break
		}
	}

	// If the wrong child subtree is not balanced then keep on searching
	if !isBalanced(graph, totalWeight, wrongChild) {
		return findCorrectedWeight(graph, weight, totalWeight, wrongChild)
	}

	// If there is a wrong weight found the return the corrected one
	diff := int(math.Abs(float64(totalWeight[wrongChild]) - float64(wantedWeight)))
	if weight[wrongChild] > diff {
		return weight[wrongChild] - diff
	}

	return weight[wrongChild] + diff
}

func run(s string) string {
	graph := make(map[string][]string)
	weight := make(map[string]int)
	for _, line := range strings.Split(s, "\n") {
		parts := strings.Split(line, " ")
		w, _ := strconv.ParseInt(parts[1][1:len(parts[1])-1], 10, 32)
		weight[parts[0]] = int(w)
		if len(parts) == 2 {
			graph[parts[0]] = []string{}
		} else {
			graph[parts[0]] = append(graph[parts[0]], strings.Split(strings.Join(parts[3:], ""), ",")...)
		}
	}

	root := findRoot(graph)
	totalWeight := calcTotalWeights(graph, weight, root)
	answer := findCorrectedWeight(graph, weight, totalWeight, root)

	return fmt.Sprint(answer)
}

func main() {
	fmt.Println(run(os.Args[1]))
}

File: day-07/part-2/ayoub.py
from runners.python import Submission


class AyoubSubmission(Submission):

	def find_root(self, graph):
		for node in graph:
			has_parent = False
			for x in graph:
				if node in graph[x]:
					has_parent = True
					break
			if not has_parent:
				return node

	def get_total_weight(self, graph, weight, root):
		if root in self._cache:
			return self._cache[root]
		s = weight[root]
		for child in graph[root]:
			s += self.get_total_weight(graph, weight, child)
		self._cache[root] = s
		return self._cache[root]

	def is_balanced(self, graph, weight, root):
		if len(graph[root]) == 0:
			return True
		return len(set([ self.get_total_weight(graph, weight, child) for child in graph[root] ])) == 1

	def find_right_weight(self, graph, weight, root):
		if len(graph[root]) == 0 or self.is_balanced(graph, weight, root):
			return None
		# Calculate total weights of children trees
		children_weights = {
			child: self.get_total_weight(graph, weight, child) for child in graph[root]
		}
		# Look for the wrong weight
		wrong_child = None
		wanted_weight = 0
		occ = {}
		for w in children_weights.values():
			if w in occ: 
				occ[w] += 1
				wanted_weight = w
			else: occ[w] = 1
		for child, w in children_weights.items():
			if occ[w] == 1:
				wrong_child = child
				break
		# If the wrong child subtree is not balanced then keep on searching 
		if not self.is_balanced(graph, weight, wrong_child):
			return self.find_right_weight(graph, weight, wrong_child)

		# If there is a wrong weight found the return the corrected one
		diff = abs(children_weights[wrong_child] - wanted_weight)
		if weight[wrong_child] > diff:
			return weight[wrong_child] - diff
		return weight[wrong_child] + diff

	def run(self, s):
		graph = {}
		weight = {}
		for line in s.split('\n'):
			parts = line.split()
			weight[parts[0]] = int(parts[1][1:-1])
			if len(parts) == 2:
				graph[parts[0]] = set()
			else:
				graph[parts[0]] = set(''.join(parts[3:]).split(','))
		
		root = self.find_root(graph)
		self._cache = {}
		return self.find_right_weight(graph, weight, root)